File: typescript/core/src/mobileSelectors.ts
const UIAUTOMATOR_SELECTOR_KEYS = new Set([
  "text",
  "textcontains",
  "textmatches",
  "textstartswith",
  "classname",
  "classnamematches",
  "description",
  "desc",
  "descriptioncontains",
  "desccontains",
  "descriptionmatches",
  "descmatches",
  "descriptionstartswith",
  "descstartswith",
  "checkable",
  "checked",
  "clickable",
  "longclickable",
  "scrollable",
  "enabled",
  "focusable",
  "focused",
  "selected",
  "packagename",
  "package",
  "packagenamematches",
  "resourceid",
  "resourceidmatches",
  "index",
  "instance",
]);

const SELECTOR_PREFIXES = ["xpath=", "xpath:", "css=", "css:", "uia=", "uia:"] as const;

type MobileSelectorFlavor = "css" | "xpath" | "uia";

function decodeSelectorBody(body: string): string {
  const candidate = body.trim();
  if (candidate.startsWith("\"") && candidate.endsWith("\"")) {
    try {
      return JSON.parse(candidate) as string;
    } catch {
      return candidate;
    }
  }
  return candidate;
}

function parseExplicitSelectorPrefix(
  selector: string,
): { flavor: MobileSelectorFlavor; prefixLength: number } | null {
  const lowered = selector.toLowerCase();
  if (lowered.startsWith("xpath=") || lowered.startsWith("xpath:")) {
    return { flavor: "xpath", prefixLength: 6 };
  }
  if (lowered.startsWith("css=") || lowered.startsWith("css:")) {
    return { flavor: "css", prefixLength: 4 };
  }
  if (lowered.startsWith("uia=") || lowered.startsWith("uia:")) {
    return { flavor: "uia", prefixLength: 4 };
  }
  return null;
}

function findJsonStringEnd(value: string): number | null {
  if (!value.startsWith("\"")) {
    return null;
  }
  let escaped = false;
  for (let index = 1; index < value.length; index += 1) {
    const char = value[index];
    if (escaped) {
      escaped = false;
      continue;
    }
    if (char === "\\") {
      escaped = true;
      continue;
    }
    if (char === "\"") {
      return index + 1;
    }
  }
  return null;
}

function isNormalizedTransportSelector(selector: string): boolean {
  const trimmed = selector.trim();
  if (!trimmed) {
    return false;
  }

  let index = 0;
  while (index < trimmed.length) {
    const prefix = parseExplicitSelectorPrefix(trimmed.slice(index));
    if (!prefix) {
      return false;
    }

    index += prefix.prefixLength;
    const remainder = trimmed.slice(index);
    if (!remainder.startsWith("\"")) {
      return false;
    }

    const jsonEnd = findJsonStringEnd(remainder);
    if (!jsonEnd) {
      return false;
    }

    index += jsonEnd;
    const tail = trimmed.slice(index);
    if (!tail) {
      return true;
    }
    if (!/^\s+/.test(tail)) {
      return false;
    }

    const nextIndex = index + tail.match(/^\s+/)?.[0].length!;
    const nextSegment = trimmed.slice(nextIndex).toLowerCase();
    if (!SELECTOR_PREFIXES.some((prefixValue) => nextSegment.startsWith(prefixValue))) {
      return false;
    }
    index = nextIndex;
  }

  return true;
}

function parseUiAutomatorSelectorPrefix(selector: string): number | null {
  const separatorIndex = selector.search(/[=:]/);
  if (separatorIndex <= 0) {
    return null;
  }
  const key = selector.slice(0, separatorIndex).trim().toLowerCase();
  if (!UIAUTOMATOR_SELECTOR_KEYS.has(key)) {
    return null;
  }
  return separatorIndex + 1;
}

function parseSelectorForTransport(selector: string): { flavor: MobileSelectorFlavor; body: string } {
  const trimmed = selector.trim();
  const explicit = parseExplicitSelectorPrefix(trimmed);
  if (explicit) {
    return { flavor: explicit.flavor, body: decodeSelectorBody(trimmed.slice(explicit.prefixLength)) };
  }
  const uiAutomatorPrefix = parseUiAutomatorSelectorPrefix(trimmed);
  if (uiAutomatorPrefix) {
    return { flavor: "uia", body: `${trimmed.slice(0, uiAutomatorPrefix - 1)}=${trimmed.slice(uiAutomatorPrefix)}` };
  }
  if (
    trimmed.startsWith("//") ||
    trimmed.startsWith(".//") ||
    trimmed.startsWith("../") ||
    trimmed.startsWith("/") ||
    trimmed.startsWith("(")
  ) {
    return { flavor: "xpath", body: trimmed };
  }
  return { flavor: "css", body: trimmed };
}

export function normalizeMobileSelectorForTransport(selector: string): string {
  const trimmed = selector.trim();
  if (!trimmed) {
    return "";
  }
  if (isNormalizedTransportSelector(trimmed)) {
    return trimmed;
  }
  const parsed = parseSelectorForTransport(selector);
  return `${parsed.flavor}=${JSON.stringify(parsed.body)}`;
}

export function chainMobileSelectorForTransport(parent: string, child: string): string {
  const parentSelector = normalizeMobileSelectorForTransport(parent);
  const childSelector = normalizeMobileSelectorForTransport(child);
  if (!parentSelector) {
    return childSelector;
  }
  if (!childSelector) {
    return parentSelector;
  }
  return `${parentSelector} ${childSelector}`;
}

File: python/allwright/_mobile_selectors.py
from __future__ import annotations

import json

UIAUTOMATOR_SELECTOR_KEYS = {
    "text",
    "textcontains",
    "textmatches",
    "textstartswith",
    "classname",
    "classnamematches",
    "description",
    "desc",
    "descriptioncontains",
    "desccontains",
    "descriptionmatches",
    "descmatches",
    "descriptionstartswith",
    "descstartswith",
    "checkable",
    "checked",
    "clickable",
    "longclickable",
    "scrollable",
    "enabled",
    "focusable",
    "focused",
    "selected",
    "packagename",
    "package",
    "packagenamematches",
    "resourceid",
    "resourceidmatches",
    "index",
    "instance",
}


def parse_explicit_selector_prefix(selector: str) -> tuple[str, int] | None:
    lower = selector.lower()
    if lower.startswith("xpath=") or lower.startswith("xpath:"):
        return "xpath", 6
    if lower.startswith("css=") or lower.startswith("css:"):
        return "css", 4
    if lower.startswith("uia=") or lower.startswith("uia:"):
        return "uia", 4
    return None


def parse_uiautomator_selector_prefix(selector: str) -> int | None:
    for index, char in enumerate(selector):
        if char not in "=:":
            continue
        key = selector[:index].strip().lower()
        if key in UIAUTOMATOR_SELECTOR_KEYS:
            return index + 1
        return None
    return None


def find_json_string_end(value: str) -> int | None:
    if not value.startswith('"'):
        return None
    escaped = False
    for index in range(1, len(value)):
        char = value[index]
        if escaped:
            escaped = False
            continue
        if char == "\\":
            escaped = True
            continue
        if char == '"':
            return index + 1
    return None


def is_normalized_transport_selector(selector: str) -> bool:
    trimmed = selector.strip()
    if not trimmed:
        return False

    index = 0
    while index < len(trimmed):
        prefix = parse_explicit_selector_prefix(trimmed[index:])
        if prefix is None:
            return False
        _, prefix_len = prefix
        index += prefix_len

        remainder = trimmed[index:]
        json_end = find_json_string_end(remainder)
        if json_end is None:
            return False
        index += json_end

        if index == len(trimmed):
            return True

        whitespace_start = index
        while index < len(trimmed) and trimmed[index].isspace():
            index += 1
        if index == whitespace_start:
            return False
        if parse_explicit_selector_prefix(trimmed[index:]) is None:
            return False

    return True


def decode_selector_body(body: str) -> str:
    candidate = body.strip()
    if len(candidate) >= 2 and candidate[0] == '"' and candidate[-1] == '"':
        try:
            decoded = json.loads(candidate)
        except json.JSONDecodeError:
            return candidate
        if isinstance(decoded, str):
            return decoded
    return candidate


def parse_selector_for_transport(selector: str) -> tuple[str, str]:
    trimmed = selector.strip()
    explicit = parse_explicit_selector_prefix(trimmed)
    if explicit is not None:
        flavor, prefix_len = explicit
        return flavor, decode_selector_body(trimmed[prefix_len:])
    uia_prefix_len = parse_uiautomator_selector_prefix(trimmed)
    if uia_prefix_len is not None:
        return "uia", f"{trimmed[:uia_prefix_len - 1]}={trimmed[uia_prefix_len:]}"
    if (
        trimmed.startswith("//")
        or trimmed.startswith(".//")
        or trimmed.startswith("../")
        or trimmed.startswith("/")
        or trimmed.startswith("(")
    ):
        return "xpath", trimmed
    return "css", trimmed


def normalize_mobile_selector_for_transport(selector: str) -> str:
    trimmed = selector.strip()
    if not trimmed:
        return ""
    if is_normalized_transport_selector(trimmed):
        return trimmed
    flavor, body = parse_selector_for_transport(selector)
    return f"{flavor}={json.dumps(body)}"


def chain_mobile_selector_for_transport(parent: str, child: str) -> str:
    parent_selector = normalize_mobile_selector_for_transport(parent) if parent.strip() else ""
    child_selector = normalize_mobile_selector_for_transport(child) if child.strip() else ""
    if not parent_selector:
        return child_selector
    if not child_selector:
        return parent_selector
    return f"{parent_selector} {child_selector}"
